sdks/java/io/synthetic/src/main/java/org/apache/beam/sdk/io/synthetic/SyntheticUnboundedSource.java

```diff
@@ -0,0 +1,215 @@
+            ? sourceOptions.forceNumInitialBundles
+            : desiredNumSplits;
+
+    List<SyntheticUnboundedSource> splits =
+        bundleSplitter
+            .getBundleSizes(desiredNumBundles, startOffset, endOffset)
+            .stream()
+            .map(
+                offsetRange ->
+                    new SyntheticUnboundedSource(
+                        offsetRange.getFrom(), offsetRange.getTo(), sourceOptions))
+            .collect(Collectors.toList());
+    LOG.info("Split into {} bundles of sizes: {}", splits.size(), splits);
+
+    return splits;
+  }
+
+  @Override
+  public boolean requiresDeduping() {
+    return true;
+  }
+
+  private class SyntheticUnboundedReader extends UnboundedReader<KV<byte[], byte[]>> {
+
+    private final SyntheticUnboundedSource source;
+
+    private KV<byte[], byte[]> currentKVPair;
+
+    private long currentOffset;
+
+    private Instant processingTime;
+
+    private Instant eventTime;
+
+    private SyntheticWatermark syntheticWatermark;
+
+    private ReaderDelay delay;
+
+    public SyntheticUnboundedReader(SyntheticUnboundedSource source) {
+      this.currentKVPair = null;
+      this.delay = new ReaderDelay(sourceOptions);
+      this.source = source;
+      this.currentOffset = 0;
+      this.syntheticWatermark = new SyntheticWatermark(sourceOptions);
+    }
+
+    @Override
+    public SyntheticUnboundedSource getCurrentSource() {
+      return source;
+    }
+
+    @Override
+    public KV<byte[], byte[]> getCurrent() throws NoSuchElementException {
+      if (currentKVPair == null) {
+        throw new NoSuchElementException(
+            "Current record is unavailable because either the reader is "
+                + "at the beginning of the input and start() or advance() wasn't called, "
+                + "or the last start() or advance() returned false.");
+      }
+      return currentKVPair;
+    }
+
+    @Override
+    public Instant getCurrentTimestamp() throws NoSuchElementException {
+      if (eventTime == null) {
+        throw new NoSuchElementException(
+            "Current timestamp is unavailable because either the reader is "
+                + "at the beginning of the input and start() or advance() wasn't called, "
+                + "or the last start() or advance() returned false.");
+      }
+      return eventTime;
+    }
+
+    @Override
+    public boolean start() {
+      currentOffset = SyntheticUnboundedSource.this.startOffset;
+      delay.delayStart(currentOffset);
+      return advance();
+    }
+
+    @Override
+    public boolean advance() {
+      currentOffset++;
+
+      processingTime = new Instant();
+      eventTime = processingTime.minus(sourceOptions.nextProcessingTimeDelay(currentOffset));
+
+      SyntheticSourceOptions.Record record =
+          getCurrentSource().sourceOptions.genRecord(currentOffset);
+      currentKVPair = record.kv;
+
+      delay.delayRecord(record);
+
+      return currentOffset < source.endOffset;
```

Comment:
I investigated today and unfortunately, this won't be enough for streaming scenarios. In order to stop the pipeline (with DONE state) after all records are generated we need to do something like Nexmark does: 

```
  if(currentOffset >= source.endOffset) {
    watermark = BoundedWindow.TIMESTAMP_MAX_VALUE;
    System.out.println("stopping the source!");
    return false;
  }  else {
   return true;
  }
```

Thanks to this the pipeline won't hang infinitely waiting for new records. I'm going to extend this PR with those changes tomorow but giving a heads up in case you want to provide some feedback about this.

Reply:
I think that is great. I was thinking that we would just cancel the pipeline from within the test code. I love that the source itself can finalize the pipeline.